File: solutions/elixir/largest-series-product/1/lib/series.ex
defmodule Series do
  @doc """
  Finds the largest product of a given number of consecutive numbers in a given string of numbers.
  """
  @spec largest_product(String.t(), non_neg_integer) :: non_neg_integer
  def largest_product(_number_string, 0),
    do: 1

  def largest_product(nil = _number_string, _size),
    do: :error

  def largest_product("" = _number_string, _size),
    do: raise(ArgumentError)

  def largest_product(number_string, size)
      when size < 0,
      do: raise(ArgumentError)

  def largest_product(number_string, size) do
    if all_characters_are_digits(number_string) &&
         String.length(number_string) >= size,
       do: find_max_product(number_string, size),
       else: raise(ArgumentError)
  end

  defp all_characters_are_digits(number_string),
    do:
      "^[+-]?[0-9]*\.?[0-9]*$"
      |> Regex.compile!()
      |> Regex.match?(number_string)

  defp find_max_product(number_string, size),
    do:
      find_products(number_string, size)
      |> Enum.max()

  defp find_products(number_string, size),
    do:
      0..(String.length(number_string) - size)
      |> Enum.map(&extract_subseries(number_string, size, &1))
      |> Enum.map(&pair_to_product/1)

  defp extract_subseries(number_string, size, start),
    do: String.slice(number_string, start, size)

  defp pair_to_product(substring),
    do:
      substring
      |> String.graphemes()
      |> Enum.map(&Integer.parse/1)
      |> Enum.map(&elem(&1, 0))
      |> IO.inspect(label: '### pair')
      |> product()

  defp product(factors),
    do: product(factors, _accumulator = 1)

  defp product([] = _factors, accumulator),
    do: accumulator

  defp product([head | tail] = _factors, accumulator),
    do: product(tail, head * accumulator)
end

File: solutions/scala/largest-series-product/1/src/main/scala/Series.scala
object Series {
  def largestProduct(size: Int, series: String): Option[Int] =
    if (size == 0) {
      Some(1)
    }
    else if (this.isInvalidInput(size, series)) {
      None
    }
    else {
      Some(this.findMaxProduct(size, series))
    }

  private def isInvalidInput(size: Int, series: String): Boolean =
    size < 0 ||
      size > series.length ||
      series.isBlank ||
      !series.forall(Character.isDigit)

  private def findMaxProduct(size: Int, series: String) = {
    this.findProducts(size, series)
      .maxBy(_._2)
      ._2
  }

  private def findProducts(size: Int, series: String): Map[String, Int] =
    (0 to (series.length - size))
      .map(extractSubseries(size, series))
      .map(pairToProduct)
      .toMap

  private def extractSubseries(size: Int, series: String)(start: Int): String =
    series.slice(start, start + size)

  private def pairToProduct(substring: String): (String, Int) = {
    val product = substring.split("")
      .map(_.toInt)
      .product
    (substring, product)
  }
}